### PR TITLE
Feat: new UI components

### PR DESCRIPTION
**Pull request by fjebaker**

### CODE REVIEW

File: src/interactive/selection.py
Comment (fjebaker) [inline src/interactive/selection.py:1]: This file needs to be renamed to reflect what it is.

File: src/interactive/selection.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import asyncio
+import logging
+
+import discord
+
+logger = logging.Logger(__name__)
```
Comment (fjebaker) [inline src/interactive/selection.py]: Logger is currently unused.

File: src/interactive/userunique.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+from interactive.selection import TimedView
+
+logger = logging.getLogger(__name__)
+
+
+class PromptModal(discord.ui.Modal, title="e-bot"):
+    def __init__(self, content):
+        super().__init__()
+        self.response = discord.ui.TextInput(label=content)
```
Comment (fjebaker) [inline src/interactive/userunique.py]: May want to try out different text input types. The default one seems to remember your inputs (at least on discord browser) which might get a bit irritating after many games. Could be the paragraph text input is more amnesiac?

File: src/abstracts/egamefactory.py [hunk 3]
```diff
@@ -23,6 +26,41 @@ def replace_rules(content: str) -> str:
...
+
+    @discord.ui.button(label="Join", style=discord.ButtonStyle.primary)
+    async def button_callback(self, interaction: discord.Interaction, button):
+        # append user
+        self.players.append(interaction.user)
+        await self._update_player_list()
+        # keep listening for more events
+        await interaction.response.defer()
+
+
```
Comment (fjebaker) [inline src/abstracts/egamefactory.py]: This should be moved to the `interactions` module.

File: src/econfig.py [hunk 4]
```diff
@@ -1,2 +1,2 @@
 PATH_EXTENSION = "."
-PLAYER_GATHER_TIMEOUT = 16
+PLAYER_GATHER_TIMEOUT = 5
```
Comment (fjebaker) [inline src/econfig.py]: Needs reverting!

File: src/abstracts/egamefactory.py [hunk 5]
```diff
@@ -189,17 +226,12 @@ async def _players_prompt(self) -> dict:
 
         :return: Authors indexed by id.
         """
-        embed = self.embed(
-            f"{self.game_description}\n\nReply to this message to join the game."
-        )
 
-        # create reply reaction pipeline
-        ipl = InteractionPipeline(ReplyInteraction())
+        text = f"{self.game_description}\n\nReply to this message to join the game."
```
Comment (JR-Mitchell) [inline src/abstracts/egamefactory.py]: This copy needs updating

File: pyproject.toml
Comment (JR-Mitchell) [inline pyproject.toml:1]: I presume this is a new type of dependencies / build running spec?

File: src/abstracts/egamefactory.py [hunk 6]
```diff
@@ -23,6 +26,41 @@ def replace_rules(content: str) -> str:
...
 
+class GatherPlayersView(TimedView):
+    def __init__(self, embed):
+        super().__init__(embed, timeout=PLAYER_GATHER_TIMEOUT)
+        self.players = []
+        self.text: str = self.embed.description
+
+    async def _update_player_list(self):
+        text = (
+            self.text + "\n\nPlayers:\n" + "\n -> ".join((i.name for i in self.players))
```
Comment (JR-Mitchell) [inline src/abstracts/egamefactory.py]: Should be:
```
self.text + "\n\nPlayers:\n-> " + "\n -> ".join((i.name for i in self.players))
```
to ensure that the first player also has `->`

File: src/games/elash.py [hunk 7]
```diff
@@ -31,13 +36,12 @@ def randomize_prompts(prompts: list) -> list:
...
 
 class ELash(EGameFactory):
-
     # configuration
 
     game_name = "E Lash"
     game_description = ""
     wait_duration = 5
-    min_players = 2
+    min_players = 1
```
Comment (JR-Mitchell) [inline src/games/elash.py]: Needs reverting too?

File: src/interactive/userunique.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    async def user_input(self, interaction: discord.Interaction, button):
+        modal = PromptModal(self.content)
+        await interaction.response.send_modal(modal)
+        # wait for response
+        await modal.wait()
+
+        self.outcome = modal.response.value
+        await self.resolve(interaction.message, modal.interaction)
+
+    async def on_timeout(self):
```
Comment (JR-Mitchell) [inline src/interactive/userunique.py:55]: What is the behaviour here if the component times out but one of the `await` statements (e.g `await modal.wait()`) is still executing?

File: pyproject.toml
Comment (fjebaker) [reply pyproject.toml:1]: https://github.com/pdm-project/pdm

File: src/games/elash.py (hunk 7 above)
Comment (fjebaker) [reply src/games/elash.py]: Yes

File: src/interactive/userunique.py (hunk 8 above)
Comment (fjebaker) [reply src/interactive/userunique.py:55]: I believe the discord library kills any hanging coros in that event loop, calls `on_timeout`, and the yields to wherever `await thing.wait()` is?

**Pull request merged**